packages/apps/contracts/EthPaymentApp.sol

```diff
@@ -0,0 +1,90 @@
+pragma solidity 0.5.8;
+pragma experimental "ABIEncoderV2";
+
+import "@counterfactual/contracts/contracts/interfaces/CounterfactualApp.sol";
+import "@counterfactual/contracts/contracts/interfaces/TwoPartyOutcome.sol";
+import "@counterfactual/contracts/contracts/interfaces/Interpreter.sol";
+import "@counterfactual/contracts/contracts/interpreters/ETHInterpreter.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+contract EthPaymentApp is CounterfactualApp {
+
+  using SafeMath for uint256;
+
+  struct AppState {
+    ETHInterpreter.ETHTransfer[2] transfers; // [sender, receiver]
+  }
+
+  struct Action {
+    uint256 paymentAmount;
+  }
+
+  // unidirectional channel, only sender can pay
+  function getTurnTaker(
+    bytes calldata encodedState, address[] calldata /* signingKeys */
+  )
+    external
+    pure
+    returns (address)
+  {
+    AppState memory state = abi.decode(encodedState, (AppState));
+    return state.transfers[0].to;
+  }
+
+  function computeOutcome(bytes calldata encodedState)
+    external
+    pure
+    returns (bytes memory)
+  {
+    return encodedState;
+  }
+
+  function applyAction(
+    bytes calldata encodedState, bytes calldata encodedAction
+  )
+    external
+    pure
+    returns (bytes memory)
+  {
+    AppState memory state = abi.decode(encodedState, (AppState));
+    Action memory action = abi.decode(encodedAction, (Action));
+
+    // apply transition based on action
+    AppState memory postState = applyPayment(state, action.paymentAmount);
+
+    return abi.encode(postState);
+  }
+
+  function applyPayment(
+    AppState memory state,
+    uint256 paymentAmount
+  )
+    internal
+    pure
+    returns (AppState memory)
+  {
+    // subtract payment amount from sender balance
+    // SafeMath will throw if below zero
+    state.transfers[0].amount = state.transfers[0].amount.sub(paymentAmount);
+    // add payment amount to receiver balance
+    state.transfers[1].amount = state.transfers[1].amount.add(paymentAmount);
+    return state;
+  }
+
+  function isStateTerminal(bytes calldata encodedState)
+    external
+    pure
+    returns (bool)
+  {
+    // always terminal
+    return true;
+  }
```

Comment:
I think you might want to restructure it such that there is a receiver "finalize payment" action where that moves it into a terminal state but whenever the sender sends funds it is not in a finalized state. The reason being is that technically right now the Node is written such that the counterparty will auto-sign the new state and return it to the sender every time. Thus, the sender in this app would have a final state for every time he sent money. So he could submit state with nonce = 1 where he only sent a small amount and it would execute the effect on chain immediately because it is terminal even though a nonce = 10 might exist.

That being said, if the Node did not auto-sign everything and the receiver simply did not return back the doubly signed copy to the sender, having every state being final would work fine in the way this contract is written already. However with one change. The actions would need to be updated in a way where the `paymentAmount` is the _total_ amount that has been sent thus far, not just the amount sent in this one payment. If you go with the approach I mention in the first paragraph though, you don't need to do this.

Reply:
https://github.com/counterfactual/monorepo/pull/1561/commits/550bd5da4436ac11ec54c887cdcdeaf74234e340

Reply:
This totally makes sense, I implemented this the way you mentioned: https://github.com/counterfactual/monorepo/pull/1561/commits/550bd5da4436ac11ec54c887cdcdeaf74234e340.

However, thinking about it more, we might not want the limitation of having to wait for a signature to be able to finalize the state on a unidirectional payment. I'll leave as is and we can revisit this.